### PR TITLE
Handle fetch errors

### PR DESCRIPTION
**Pull request by aistrate**

### CODE REVIEW

File: src/coffeeShops/CoffeeShopsView.tsx [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+      {coffeeShopsFetch.data &&
+        coffeeShopsFetch.data.map((coffeeShop) => (
+          <CoffeeShopView key={coffeeShop.id} coffeeShop={coffeeShop} />
+        ))}
+    </div>
+  );
+}
+
+function CoffeeShopView({ coffeeShop }: { coffeeShop: CoffeeShop }) {
```
Comment (dasilvaluis) [inline src/coffeeShops/CoffeeShopsView.tsx]: Would it make sense to have this a separate component?

File: src/coffeeShops/useCoffeeShopsFetch.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+  if (
+    fetchResult.httpStatus === 200 ||
+    (!fetchResult.httpStatus && !fetchResult.timedOut) ||
+    retries <= 0
+  ) {
+    return fetchResult;
+  }
+
+  if (fetchResult.httpStatus !== 401 && !fetchResult.timedOut) {
```
Comment (dasilvaluis) [inline src/coffeeShops/useCoffeeShopsFetch.ts]: I feel that the approach of using a plain number to assess http status might make it harder for the reader to understand what to expect, vs using a constant. What do you think?

File: src/coffeeShops/useCoffeeShopsFetch.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+type CoffeeShopsFetchResult = { token?: string } & FetchResult<CoffeeShop[]>;
+
+export default function useCoffeeShopsFetch() {
+  const [fetchResult, setFetchResult] = useState<CoffeeShopsFetchResult>({});
+  const [isLoading, setIsLoading] = useState(true);
+
+  useEffect(() => {
+    (async () => {
+      setFetchResult(await fetchCoffeeShopsWithRetry(retryCount));
```
Comment (dasilvaluis) [inline src/coffeeShops/useCoffeeShopsFetch.ts]: There is quite a bit of boilerplate going on just to have an async operation inside the effect. Can we just have a separate function?

File: src/coffeeShops/useCoffeeShopsFetch.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+async function fetchCoffeeShopsWithRetry(
+  retries: number,
+  prevToken?: string
+): Promise<CoffeeShopsFetchResult> {
+  const fetchResult = await fetchCoffeeShops(prevToken);
+
+  retries = retries - 1;
+
+  if (
+    fetchResult.httpStatus === 200 ||
```
Comment (dasilvaluis) [inline src/coffeeShops/useCoffeeShopsFetch.ts]: These condition grew a bit too much that it becomes hard to follow the full thing. Would it be nicer to have const with meaningful name?

**Review by dasilvaluis — CHANGES_REQUESTED** [review]

File: src/coffeeShops/useCoffeeShopsFetch.ts (hunk 2 above)
Comment (aistrate) [reply src/coffeeShops/useCoffeeShopsFetch.ts]: I agree. I will change it.

File: src/coffeeShops/CoffeeShopsView.tsx (hunk 1 above)
Comment (aistrate) [reply src/coffeeShops/CoffeeShopsView.tsx]: Do you mean in a separate file? Anyway, this is temporary code, the idea is to show the shops/coordinates in a 2D grid, using a charting library.

File: src/coffeeShops/useCoffeeShopsFetch.ts (hunk 3 above)
Comment (aistrate) [reply src/coffeeShops/useCoffeeShopsFetch.ts]: Will change it.

File: src/coffeeShops/useCoffeeShopsFetch.ts (hunk 4 above)
Comment (aistrate) [reply src/coffeeShops/useCoffeeShopsFetch.ts]: I agree.

File: src/coffeeShops/CoffeeShopsView.tsx (hunk 1 above)
Comment (dasilvaluis) [reply src/coffeeShops/CoffeeShopsView.tsx]: Yep, that would be the case. Ok, got it

File: src/coffeeShops/useCoffeeShopsFetch.ts (hunk 2 above)
Comment (adrianbunea) [reply src/coffeeShops/useCoffeeShopsFetch.ts]: My personal opinion is that 401 is a no-brainer, I see no value in extracting it, the lane is perfectly readable too.
I think extracting constants is useful for arbitrary values / our own business logic / less known standards. Not staples of HTTP like Status Codes.

Comment (aistrate) [reply src/coffeeShops/useCoffeeShopsFetch.ts]: Adrian, I hope my solution doesn't make code harder to read. :)

Comment (adrianbunea) [reply src/coffeeShops/useCoffeeShopsFetch.ts]: It's fine, I realize it is a matter of taste.

**Review by adrianbunea — APPROVED** [review]

**Review by dasilvaluis — APPROVED** [review]
